network/src/lib.rs

```diff
@@ -3,61 +3,129 @@ pub mod propagation_network;
-pub struct BroadcastStatus {
-    relayed_nodes: Vec<PublicKey>,
+pub struct NetworkConfig {
+    /// The set of the members of the network.
+    pub members: Vec<PublicKey>,
+    /// The public key of this node.
+    pub public_key: PublicKey,
+    /// The private key of this node.
+    pub private_key: PrivateKey,
 }
 
-/// TODO: Provide error types.
-///
-/// Note: This trait is quite subject to change.
+/// The p2p gossip network.
 #[async_trait]
-pub trait AuthorizedNetwork: Send + Sync {
-    /// Joins the network with an authorized identity.
-    async fn new(
-        public_key: PublicKey,
-        private_key: PrivateKey,
-        known_peers: Vec<PublicKey>,
-        bootstrap_points: Vec<SocketAddrV4>,
-        network_id: String,
-    ) -> Result<Self, String>
-    where
-        Self: Sized;
-    /// Broadcasts a message to the network, after signed by the key given to this instance.
-    async fn broadcast(&self, message: &[u8]) -> Result<BroadcastToken, String>;
-    /// Stops a currently broadcasting message.
-    async fn stop_broadcast(&self, token: BroadcastToken) -> Result<(), String>;
-    /// Gets the current status of a broadcasting message.
-    async fn get_broadcast_status(&self, token: BroadcastToken) -> Result<BroadcastStatus, String>;
-    /// Creates a receiver for every message broadcasted to the network, except the one sent by this instance.
-    async fn create_recv_queue(&self) -> Result<broadcast::Receiver<Vec<u8>>, ()>;
-    /// Provides the estimated list of live nodes that are eligible and identified by their public keys.
-    async fn get_live_list(&self) -> Result<Vec<PublicKey>, ()>;
-}
-
-/// TODO: Provide error types.
+pub trait P2PNetwork {
+    /// Broadcasts a message to the network.
+    async fn broadcast(
+        config: &NetworkConfig,
+        known_peers: &[Peer],
+        message: Vec<u8>,
+    ) -> Result<(), Error>;
+
+    /// Remains online on the network indefinitely,
+    /// relaying (propagating) messages broadcasted over the network.
+    ///
+    /// * `send` - A channel to send the received messages to.
+    async fn serve(
+        config: &NetworkConfig,
+        peers: Arc<RwLock<Vec<Peer>>>,
+        send: mpsc::Sender<Vec<u8>>,
+    ) -> Result<(), Error>;
+}
+
+/// A **cumulative** set that is shared in the p2p network, backed by the local file system.
+///
+/// One of the notable characteristics of blockchain is that it is based on heights;
+/// The key idea here is that we retanin an instance (both in memory or on disk)
+/// of `DistributedMessageSet` only for a specific height,
+/// and discard if the height progresses, creating a new and empty one again.
 ///
-/// Note: This trait is quite subject to change.
+/// For every methods,
+/// - If the given directory is empty, it fails (except `create()`).
+/// - It locks the storage.
+/// - If the given directory is locked (possibly by another instnace of `DistributedMessageSet`),
+/// it blocks until the lock is released.
```

Comment:
```suggestion
/// it will `await` on the lock until it is released.
```
Possibly misleading term?